Sources/ClientRuntime/Interceptor/Interceptors.swift

```diff
@@ -0,0 +1,448 @@
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterDeserialization(
+        _ interceptorFn: @escaping (
+            any AfterDeserialization<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        )
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterDeserialization: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addModifyBeforeAttemptCompletion(
+        _ interceptorFn: @escaping (
+            any MutableOutputAfterAttempt<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        )
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(modifyBeforeAttemptCompletion: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterAttempt(
+        _ interceptorFn: @escaping (any AfterAttempt<InputType, OutputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterAttempt: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addModifyBeforeCompletion(
+        _ interceptorFn: @escaping (
+            any MutableOutputFinalization<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        ) async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(modifyBeforeCompletion: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterExecution(
+        _ interceptorFn: @escaping (any Finalization<InputType, OutputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterExecution: interceptorFn))
+    }
+}
+
+/// This extension is for the convenience of Orchestrator to call the same hook on all interceptors.
+/// Note that it doesn't conform to Interceptor because it is backed by AnyInterceptor, which also
+/// doesn't conform to Interceptor.
+///
+/// These methods throw the last error that occurred when executing a given hook, so if multiple
+/// interceptors fail, only the last error is thrown. The rest are logged (TODO)
+extension Interceptors {
+    internal func readBeforeExecution(context: InterceptorType.InterceptorContextType) async throws {
+        var error: Error?
+        for i in interceptors {
+            do {
+                try await i.readBeforeExecution(context: context)
+            } catch let e {
+                error = e
+            }
+        }
+        if let error = error {
+            throw error
+        }
```

Comment:
Lines 184-194 are repeated in each of these funcs with only one line differing, can we create a higher-order function that takes the relevant function as input? Should reduce code duplication and shorten this file by a lot

Ex. I didn't test this but maybe something like the following (also note potential suggestion on accomplishing logging)
```
internal func executeInterceptors<T>( // can make the name better
    context: InterceptorType.InterceptorContextType, 
    operation: @escaping (InterceptorType) async throws -> T //
) async throws {
    var error: Error? // suggestion: you could change this to a list 
    for i in interceptors {
        do {
            try await operation(i)
        } catch let e {
            error = e // append to list
        }
    }
   if let error = error {
        throw error // loop through list of errors and log the errors then throw the last one
   }
}
```

then would modify these functions like so:

```
internal func readBeforeExecution(context: InterceptorType.InterceptorContextType) async throws {
    try await executeInterceptors(context: context) { interceptor in
        try await interceptor.readBeforeExecution(context: context)
    }
}
```

Reply:
Done